File: ClipmanIOS/ClipmanIOS/Services/SharedFolderStorageClient.swift
import CryptoKit
import Foundation

enum SharedFolderStorageError: Error, LocalizedError {
    case unavailable
    case staleBookmark
    case invalidPassword
    case invalidRules

    var errorDescription: String? {
        switch self {
        case .unavailable:
            "The selected shared folder is unavailable. Clipman will keep the change locally and retry."
        case .staleBookmark:
            "Permission for the shared folder has expired. Choose the folder again in Settings."
        case .invalidPassword:
            "Shared folder sync requires a nonblank history password."
        case .invalidRules:
            "The shared sync-rules file could not be decoded."
        }
    }
}

/// Presents a coordinated file as the same revisioned storage contract used by
/// Clipman Server. The repository therefore applies one merge algorithm to both
/// transports, including deletion tombstones and channel routing.
final class SharedFolderStorageClient: @unchecked Sendable, HistoryStorageClient {
    enum ContentKind: Sendable {
        case history
        case rules
    }

    let isConfigured: Bool
    let syncCacheIdentity: String
    let databaseID: String
    let storageName = "shared folder"
    let createOnlyWhenMissing = true

    private let bookmark: Data?
    private let directDirectory: URL?
    private let password: String
    private let fileName: String
    private let contentKind: ContentKind

    init(bookmark: Data, password: String) {
        self.bookmark = bookmark
        self.directDirectory = nil
        self.password = password
        self.fileName = "clipman-history.clipdb"
        self.contentKind = .history
        self.databaseID = fileName
        self.isConfigured = !bookmark.isEmpty && !password.isEmpty
        self.syncCacheIdentity = "shared-folder|\(Self.bookmarkIdentity(bookmark))|\(fileName)"
    }

    init(directory: URL, password: String, fileName: String = "clipman-history.clipdb", contentKind: ContentKind = .history) {
        self.bookmark = nil
        self.directDirectory = directory
        self.password = password
        self.fileName = fileName
        self.contentKind = contentKind
        self.databaseID = fileName
        self.isConfigured = !password.isEmpty
        self.syncCacheIdentity = "shared-folder|\(directory.standardizedFileURL.path)|\(fileName)"
    }

    private init(parent: SharedFolderStorageClient, fileName: String, contentKind: ContentKind) {
        self.bookmark = parent.bookmark
        self.directDirectory = parent.directDirectory
        self.password = parent.password
        self.fileName = fileName
        self.contentKind = contentKind
        self.databaseID = fileName
        self.isConfigured = parent.isConfigured
        self.syncCacheIdentity = "\(parent.syncCacheIdentity.components(separatedBy: "|").dropLast().joined(separator: "|"))|\(fileName)"
    }

    func historyChannel(_ channelKey: String, password: String) -> (any HistoryStorageClient)? {
        guard isConfigured, self.password == password else { return nil }
        return SharedFolderStorageClient(
            parent: self,
            fileName: SyncRuleEngine.channelFileName(channelKey),
            contentKind: .history
        )
    }

    func historySyncRules(password: String) -> (any HistoryStorageClient)? {
        guard isConfigured, self.password == password else { return nil }
        return SharedFolderStorageClient(
            parent: self,
            fileName: SyncRuleEngine.syncRulesFileName,
            contentKind: .rules
        )
    }

    func metadata() async throws -> ServerDatabaseMetadata {
        let candidates = try readCandidates()
        guard !candidates.isEmpty else { throw ServerStorageError.notFound }
        return ServerDatabaseMetadata(revision: Self.revision(for: candidates))
    }

    func download() async throws -> ServerDatabaseDownload {
        let candidates = try readCandidates()
        guard let first = candidates.first else { throw ServerStorageError.notFound }
        let revision = Self.revision(for: candidates)
        guard candidates.count > 1 else {
            return ServerDatabaseDownload(revision: revision, data: first.data)
        }

        let merged = try await merge(candidates.map(\.data))
        return ServerDatabaseDownload(revision: revision, data: merged)
    }

    func upload(data: Data, expectedRevision: String, createOnly: Bool = false) async throws -> String {
        guard isConfigured else {
            throw password.isEmpty ? SharedFolderStorageError.invalidPassword : SharedFolderStorageError.unavailable
        }
        guard data.count <= ClipDatabaseFile.maximumFileBytes else { throw ServerStorageError.responseTooLarge }
        return try withDirectory { directory in
            let target = directory.appendingPathComponent(fileName, isDirectory: false)
            var coordinationError: NSError?
            var result: Result<String, Error>!
            var didWrite = false
            NSFileCoordinator().coordinate(writingItemAt: target, options: [], error: &coordinationError) { coordinatedURL in
                do {
                    let candidates = try Self.readCandidates(canonical: coordinatedURL)
                    if createOnly && !candidates.isEmpty { throw ServerStorageError.conflict }
                    if !expectedRevision.isEmpty,
                       Self.revision(for: candidates) != expectedRevision {
                        throw ServerStorageError.conflict
                    }
                    try data.write(to: coordinatedURL, options: [.atomic, .completeFileProtectionUnlessOpen])
                    didWrite = true
                    result = .success(Self.revision(for: [Candidate(url: coordinatedURL, data: data)]))
                } catch {
                    result = .failure(error)
                }
            }
            if let coordinationError { throw coordinationError }
            if didWrite { Self.resolveConflicts(canonical: target) }
            return try result.get()
        }
    }

    private func readCandidates() throws -> [Candidate] {
        guard isConfigured else {
            throw password.isEmpty ? SharedFolderStorageError.invalidPassword : SharedFolderStorageError.unavailable
        }
        return try withDirectory { directory in
            let target = directory.appendingPathComponent(fileName, isDirectory: false)
            var coordinationError: NSError?
            var result: Result<[Candidate], Error>!
            NSFileCoordinator().coordinate(readingItemAt: target, options: .withoutChanges, error: &coordinationError) { coordinatedURL in
                do {
                    result = .success(try Self.readCandidates(canonical: coordinatedURL))
                } catch {
                    result = .failure(error)
                }
            }
            if let coordinationError { throw coordinationError }
            return try result.get()
        }
    }

    private func merge(_ payloads: [Data]) async throws -> Data {
        let password = password
        let kind = contentKind
        return try await Task.detached(priority: .userInitiated) {
            switch kind {
            case .history:
                var merged = ClipDatabase()
                for data in payloads {
                    merged = SyncConflictResolver.merge(
                        target: merged,
                        source: try ClipDatabaseFile.load(data, password: password)
                    )
                }
                return try ClipDatabaseFile.save(
                    merged,
                    password: password,
                    preferredSalt: payloads.compactMap(ClipDatabaseFile.encryptedSalt).first
                )
            case .rules:
                var merged: SyncRulesDocument?
                for data in payloads {
                    let raw = try ClipDatabaseFile.loadRawPayload(data, password: password)
                    guard let document = SyncRuleEngine.parse(raw) else { continue }
                    merged = SyncRuleEngine.merge(local: merged, remote: document) ?? document
                }
                guard let merged, let raw = SyncRuleEngine.serialize(merged) else {
                    throw SharedFolderStorageError.invalidRules
                }
                return try ClipDatabaseFile.saveRawPayload(
                    raw,
                    password: password,
                    preferredSalt: payloads.compactMap(ClipDatabaseFile.encryptedSalt).first
                )
            }
        }.value
    }

    private func withDirectory<T>(_ body: (URL) throws -> T) throws -> T {
        let directory: URL
        if let directDirectory {
            directory = directDirectory
        } else if let bookmark {
            var stale = false
            directory = try URL(
                resolvingBookmarkData: bookmark,
                options: [],
                relativeTo: nil,
                bookmarkDataIsStale: &stale
            )
            guard !stale else { throw SharedFolderStorageError.staleBookmark }
        } else {
            throw SharedFolderStorageError.unavailable
        }
        let scoped = directDirectory == nil && directory.startAccessingSecurityScopedResource()
        guard directDirectory != nil || scoped else { throw SharedFolderStorageError.unavailable }
        defer { if scoped { directory.stopAccessingSecurityScopedResource() } }
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return try body(directory)
    }

    private struct Candidate {
        var url: URL
        var data: Data
    }

    private static func readCandidates(canonical: URL) throws -> [Candidate] {
        var urls: [URL] = []
        if FileManager.default.fileExists(atPath: canonical.path) { urls.append(canonical) }
        urls.append(contentsOf: conflictSiblings(for: canonical))
        urls.append(contentsOf: (NSFileVersion.unresolvedConflictVersionsOfItem(at: canonical) ?? []).compactMap(\.url))

        var seen = Set<String>()
        var result: [Candidate] = []
        for url in urls where seen.insert(url.standardizedFileURL.path).inserted {
            do {
                if (try? url.resourceValues(forKeys: [.isUbiquitousItemKey]).isUbiquitousItem) == true {
                    try? FileManager.default.startDownloadingUbiquitousItem(at: url)
                }
                let data = try ClipDatabaseFile.readBounded(from: url)
                result.append(Candidate(url: url, data: data))
            } catch ClipDatabaseError.databaseFileTooLarge {
                throw ServerStorageError.responseTooLarge
            } catch {
                // A cloud placeholder can exist before its bytes are available.
                // Treat that as unavailable rather than missing so a local create
                // can never replace an unhydrated history or conflict version.
                throw error
            }
        }
        return result
    }

    private static func revision(for candidates: [Candidate]) -> String {
        let hashes = candidates.map { SHA256.hash(data: $0.data).map { String(format: "%02x", $0) }.joined() }.sorted()
        let digest = SHA256.hash(data: Data(hashes.joined(separator: "|").utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    private static func resolveConflicts(canonical: URL) {
        for sibling in conflictSiblings(for: canonical) {
            try? FileManager.default.removeItem(at: sibling)
        }
        for version in NSFileVersion.unresolvedConflictVersionsOfItem(at: canonical) ?? [] {
            version.isResolved = true
        }
        try? NSFileVersion.removeOtherVersionsOfItem(at: canonical)
    }

    private static func conflictSiblings(for canonical: URL) -> [URL] {
        let directory = canonical.deletingLastPathComponent()
        let base = canonical.deletingPathExtension().lastPathComponent
        let ext = canonical.pathExtension
        guard let children = try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil) else {
            return []
        }
        return children.filter { url in
            guard url != canonical,
                  url.pathExtension.caseInsensitiveCompare(ext) == .orderedSame else { return false }
            let name = url.deletingPathExtension().lastPathComponent
            guard name.range(of: base, options: [.caseInsensitive, .anchored]) != nil else { return false }
            let suffix = name.dropFirst(base.count).lowercased()
            if suffix.contains("conflicted copy") || suffix.contains("[conflict]") || suffix.contains(" conflict") {
                return true
            }
            if suffix.hasPrefix("_conf(") || suffix.hasPrefix(" _conf(") { return true }
            return looksLikeProviderComputerSuffix(String(name.dropFirst(base.count)))
        }
    }

    private static func looksLikeProviderComputerSuffix(_ suffix: String) -> Bool {
        var value = suffix.trimmingCharacters(in: .whitespacesAndNewlines)
        guard value.count >= 3 && value.count <= 80 else { return false }
        guard value.hasPrefix("-") || value.hasPrefix("(") || value.hasPrefix(" ") else { return false }
        value = value.trimmingCharacters(in: CharacterSet(charactersIn: " -()"))
        guard value.count >= 2 && value.count <= 64 else { return false }
        return value.allSatisfy { $0.isLetter || $0.isNumber || $0 == "-" || $0 == "_" }
    }

    private static func bookmarkIdentity(_ bookmark: Data) -> String {
        SHA256.hash(data: bookmark).prefix(12).map { String(format: "%02x", $0) }.joined()
    }
}

File: ClipmanAndroid/app/src/main/java/me/onj/clipman/SharedFolderStorageClient.kt
package me.onj.clipman

import android.content.Context
import android.net.Uri
import androidx.documentfile.provider.DocumentFile
import java.security.MessageDigest

internal class SharedFolderStorageException(message: String, cause: Throwable? = null) :
    Exception(message, cause)

internal enum class SharedFolderContentKind {
    History,
    Rules
}

internal data class SharedFolderDocument(
    val id: String,
    val name: String
)

internal interface SharedFolderDocumentBackend {
    val identity: String
    fun list(): List<SharedFolderDocument>
    fun read(document: SharedFolderDocument): ByteArray
    fun write(fileName: String, data: ByteArray)
    fun delete(document: SharedFolderDocument)
}

private class AndroidSharedFolderDocumentBackend(
    context: Context,
    private val treeUri: Uri
) : SharedFolderDocumentBackend {
    private val appContext = context.applicationContext
    override val identity: String = treeUri.toString()

    override fun list(): List<SharedFolderDocument> = tree().listFiles().mapNotNull { document ->
        val name = document.name ?: return@mapNotNull null
        SharedFolderDocument(document.uri.toString(), name)
    }

    override fun read(document: SharedFolderDocument): ByteArray =
        appContext.contentResolver.openInputStream(Uri.parse(document.id))?.use { input ->
            ClipDatabaseFile.readDatabaseBlob(input)
        } ?: throw SharedFolderStorageException(
            "The selected shared-folder file could not be opened for reading."
        )

    override fun write(fileName: String, data: ByteArray) {
        val folder = tree()
        val target = folder.listFiles().firstOrNull { it.name.equals(fileName, ignoreCase = true) }
            ?: folder.createFile(mimeType, fileName)
            ?: throw SharedFolderStorageException("The shared-folder file could not be created.")
        val resolver = appContext.contentResolver
        val output = try {
            resolver.openOutputStream(target.uri, "rwt")
                ?: resolver.openOutputStream(target.uri, "wt")
        } catch (error: Throwable) {
            if (error is SecurityException) throw error
            resolver.openOutputStream(target.uri, "wt")
        } ?: throw SharedFolderStorageException(
            "The selected shared-folder file could not be opened for writing."
        )
        output.use {
            it.write(data)
            it.flush()
        }
    }

    override fun delete(document: SharedFolderDocument) {
        DocumentFile.fromSingleUri(appContext, Uri.parse(document.id))?.delete()
    }

    private fun tree(): DocumentFile {
        val folder = DocumentFile.fromTreeUri(appContext, treeUri)
            ?: throw SharedFolderStorageException("The selected shared folder is no longer available.")
        if (!folder.canRead() || !folder.canWrite()) {
            throw SharedFolderStorageException(
                "Clipman no longer has read and write permission for the selected shared folder."
            )
        }
        return folder
    }

    companion object {
        private const val mimeType = "application/octet-stream"
    }
}

/**
 * Presents a Storage Access Framework folder through the same revisioned
 * contract used by Clipman Server. Provider-specific networking and accounts
 * remain the responsibility of Android and the selected document provider.
 */
internal class SharedFolderStorageClient private constructor(
    private val backend: SharedFolderDocumentBackend,
    private val password: String,
    private val fileName: String,
    private val contentKind: SharedFolderContentKind
) : HistoryStorageClient {
    constructor(context: Context, treeUri: String, password: String) : this(
        AndroidSharedFolderDocumentBackend(context, Uri.parse(treeUri)),
        password,
        LocalHistoryStore.coreFileName,
        SharedFolderContentKind.History
    )

    internal constructor(
        backend: SharedFolderDocumentBackend,
        password: String
    ) : this(backend, password, LocalHistoryStore.coreFileName, SharedFolderContentKind.History)

    override val isConfigured = backend.identity.isNotBlank() && password.isNotBlank()
    override val storageName = "shared folder"
    override val syncCacheIdentity: String
        get() = sha256Hex("shared-folder|${backend.identity}|$fileName|${sha256Hex(password)}")
    override val createOnlyWhenMissing = true

    override fun forChannel(channelKey: String): HistoryStorageClient? {
        if (!isConfigured || channelKey.isBlank()) return null
        return SharedFolderStorageClient(
            backend,
            password,
            LocalHistoryStore.channelFileName(channelKey),
            SharedFolderContentKind.History
        )
    }

    override fun forSyncRules(): HistoryStorageClient? {
        if (!isConfigured) return null
        return SharedFolderStorageClient(
            backend,
            password,
            syncRulesFileName,
            SharedFolderContentKind.Rules
        )
    }

    override fun metadata(): String = synchronized(ioLock) {
        val candidates = readCandidates()
        if (candidates.isEmpty()) "" else revision(candidates)
    }

    override fun download(): ServerDatabaseDownload = synchronized(ioLock) {
        val candidates = readCandidates()
        if (candidates.isEmpty()) {
            throw ServerDatabaseNotFoundException("The shared-folder database does not exist yet.")
        }
        val data = if (candidates.size == 1) candidates.first().data else merge(candidates.map { it.data })
        ServerDatabaseDownload(revision(candidates), data)
    }

    override fun upload(
        data: ByteArray,
        expectedRevision: String,
        createOnly: Boolean
    ): ServerDatabaseDownload = synchronized(ioLock) {
        requireConfigured()
        ClipDatabaseFile.requireDatabaseBlobSize(data.size.toLong())
        val candidates = readCandidates()
        if (createOnly && candidates.isNotEmpty()) {
            throw ServerConflictException("The shared-folder database was created by another device.")
        }
        if (expectedRevision.isNotBlank() && revision(candidates) != expectedRevision) {
            throw ServerConflictException("The shared-folder database changed on another device.")
        }
        backend.write(fileName, data)
        candidates
            .filterNot { it.document.name.equals(fileName, ignoreCase = true) }
            .forEach { runCatching { backend.delete(it.document) } }
        ServerDatabaseDownload(revision(listOf(Candidate(SharedFolderDocument(fileName, fileName), data))), ByteArray(0))
    }

    private fun readCandidates(): List<Candidate> {
        requireConfigured()
        return backend.list()
            .filter { document ->
                document.name.equals(fileName, ignoreCase = true) ||
                    isConflictSibling(document.name, fileName)
            }
            .map { document -> Candidate(document, backend.read(document)) }
    }

    private fun merge(payloads: List<ByteArray>): ByteArray = when (contentKind) {
        SharedFolderContentKind.History -> {
            var merged = ClipDatabase()
            payloads.forEach { data ->
                merged = SyncConflictResolver.merge(merged, ClipDatabaseFile.load(data, password))
            }
            ClipDatabaseFile.save(
                merged,
                password,
                preferredSalt = payloads.firstNotNullOfOrNull(ClipDatabaseFile::encryptedSalt)
            )
        }
        SharedFolderContentKind.Rules -> {
            var merged: SyncRulesDocument? = null
            payloads.forEach { data ->
                val document = SyncRuleEngine.parse(ClipDatabaseFile.loadRawText(data, password))
                    ?: throw SharedFolderStorageException("The shared sync-rules file could not be decoded.")
                merged = SyncRuleEngine.mergeDocuments(merged, document)
            }
            val document = merged
                ?: throw SharedFolderStorageException("The shared sync-rules file could not be decoded.")
            ClipDatabaseFile.saveRawText(
                SyncRuleEngine.serialize(document),
                password,
                preferredSalt = payloads.firstNotNullOfOrNull(ClipDatabaseFile::encryptedSalt)
            )
        }
    }

    private fun requireConfigured() {
        if (password.isBlank()) {
            throw SharedFolderStorageException("Shared folder sync requires a nonblank history password.")
        }
        if (backend.identity.isBlank()) {
            throw SharedFolderStorageException("Choose a shared folder in Settings.")
        }
    }

    private data class Candidate(
        val document: SharedFolderDocument,
        val data: ByteArray
    )

    companion object {
        private const val syncRulesFileName = "clipman-sync-rules.clipdb"
        private val ioLock = Any()

        private fun revision(candidates: List<Candidate>): String {
            val hashes = candidates.map { sha256Hex(it.data) }.sorted()
            return sha256Hex(hashes.joinToString("|"))
        }

        private fun sha256Hex(value: String): String = sha256Hex(value.toByteArray(Charsets.UTF_8))

        private fun sha256Hex(value: ByteArray): String =
            MessageDigest.getInstance("SHA-256")
                .digest(value)
                .joinToString("") { "%02x".format(it) }

        internal fun isConflictSibling(candidate: String, canonical: String): Boolean {
            val extension = canonical.substringAfterLast('.', "")
            val base = canonical.removeSuffix(if (extension.isEmpty()) "" else ".$extension")
            if (!candidate.endsWith(if (extension.isEmpty()) "" else ".$extension", ignoreCase = true)) return false
            val candidateBase = candidate.removeSuffix(if (extension.isEmpty()) "" else ".${candidate.substringAfterLast('.')}")
            if (!candidateBase.startsWith(base, ignoreCase = true) || candidateBase.equals(base, ignoreCase = true)) {
                return false
            }
            val suffix = candidateBase.drop(base.length).lowercase()
            if (suffix.contains("conflicted copy") || suffix.contains("[conflict]") || suffix.contains(" conflict")) {
                return true
            }
            if (suffix.startsWith("_conf(") || suffix.startsWith(" _conf(")) return true
            return suffix.matches(Regex("""\s*[\[(]\d+[\])]\s*"""))
        }
    }
}
